src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,2991 @@
+    else
+      r.SF = parent(s[1])
+    end
+    r.S = s
+    return r
+  end
+end
+
+# ModuleGens from an Array of Oscar free module elements, specifying the free module 
+# and Singular free module, only useful indirectly
+ModuleGens(O::Vector{<:FreeModuleElem}, F::FreeMod{T}, SF::Singular.FreeMod) where T = ModuleGens{T}(O, F, SF)
+
+# ModuleGens from a Singular submodule
+ModuleGens(F::FreeMod{S}, s::Singular.smodule) where {S} = ModuleGens{S}(F, s)
+
+# ModuleGens from an Array of Oscar free module elements
+# Todo: Empty generating set
+function ModuleGens(O::Vector{<:FreeModuleElem})
+  @assert length(O) > 0
+  SF = singular_module(parent(O[1]))
+  return ModuleGens(O, SF)
+end
+
+# ModuleGens from an Array of Oscar free module elements, specifying the Oscar free module
+# note that the array might be empty
+function ModuleGens(O::Vector{<:FreeModuleElem}, F::FreeMod{T}) where {T}
+  SF = singular_module(F)
+  return ModuleGens{T}(O, F, SF)
+end
+
+# ModuleGens from an Array of Oscar free module elements, specifying the Singular free module
+# note that the array might be empty
+function ModuleGens(O::Vector{<:FreeModuleElem}, SF::Singular.FreeMod)
+  return ModuleGens{elem_type(base_ring(parent(O[1])))}(O, parent(O[1]), SF)
+end
+
+# getproperty is used to fill in fields with duplicate data
+#=function Base.getproperty(M::ModuleGens, s::Symbol)
+  if s == :S
+    error("deprecated")
+    singular_assure(M)
+    return getfield(M, s)
+  elseif s == :O
+    error("deprecated")
+    oscar_assure(M)
+    return getfield(M, s)
+  else
+    return getfield(M,s)
+  end
+end=#
+
+function singular_generators(M::ModuleGens)
+  singular_assure(M)
+  return M.S
+end
+
+function oscar_generators(M::ModuleGens)
+  oscar_assure(M)
+  return M.O
+end
+
+# Test for zero module
+function iszero(M::ModuleGens)
+  return iszero(singular_generators(M))
+end
+
+# remove output saying defined on the Singular side?
+function show(io::IO, F::ModuleGens)
+  println(io, "Array of length ", length(F))
+  for i=1:length(F)
+    if isassigned(F.O, i)
+      println(io, i, " -> ", F.O[i])
+    end
+  end
+  if isdefined(F, :S)
+    println(io, "defined on the Singular side")
+  end
+end
+
+# number of elements of the module generating set
+length(F::ModuleGens) = length(oscar_generators(F))
+# number of elements of the module generating set
+ngens(F::ModuleGens) = length(oscar_generators(F))
+
+# i-th entry of module generating set on Oscar side
+# Todo: clean up, convert or assure
+function getindex(F::ModuleGens, ::Val{:O}, i::Int)
+  if !isassigned(F.O, i)
+    F.O[i] = convert(F.F, singular_generators(F)[i])
+  end
+  return oscar_generators(F)[i]
+end
+
+# i-th entry of module generating set on Singular side
+# Todo: clean up, convert or assure
+function getindex(F::ModuleGens, ::Val{:S}, i::Int)
+  if !isdefined(F, :S)
+    F.S = Singular.smodule{elem_type(base_ring(F.SF))}(base_ring(F.SF), [convert(F.SF,x) for x = oscar_generators(F)]...)
```

Comment:
I recommend using `Singular.Module` instead (and also elsewhere in this file; and also in the rest of Oscar), which is slightly higher level and IMHO results in slightly easier to read code
```suggestion
    F.S = Singular.Module(base_ring(F.SF), [convert(F.SF,x) for x = oscar_generators(F)]...)
```